src/Controller/Controller.php

```diff
@@ -509,10 +508,20 @@ public function invokeAction(): ?ResponseInterface
             ]);
         }
 
-        /** @var callable $callable */
-        $callable = [$this, $request->getParam('action')];
+        return Closure::fromCallable([$this, $action]);
+    }
 
-        $result = $callable(...array_values($request->getParam('pass')));
+    /**
+     * Dispatches the controller action.
+     *
+     * @param \Closure $action The action closure.
+     * @param array $args The arguments to be passed when invoking action.
+     * @return void
+     * @throws \UnexpectedValueException If return value of action is not `null` or `ResponseInterface` instance.
+     */
+    public function invokeAction(Closure $action, array $args): void
+    {
+        $result = $action(...$args);
```

Comment:
Are you thinking that a DI container could be another closure that wraps the action's closure and performs injection?

Reply:
I was going by what's done here https://github.com/robotusers/cakephp-di/blob/master/src/Http/ControllerFactory.php#L106-L134. It manually goes through the arguments list of the closure and generates list of args to pass.

Happy to change if there's a better way to invoke the closure with required args.

Reply:
This will work well. Building a new argument list with the injected parameters is simpler than what I was thinking.